File: pset6/credit/credit.py
from cs50 import get_int, get_string

# get user input as string
number = get_string("Number: ")

# Create list of integers
l = [int(n) for n in number]

# Create variable for sum
s = 0

# Every other digit starting with the second-to-last
for n in range(len(l) - 2, -1, -2):
    if 2 * l[n] > 9:
        s += int(str(l[n] * 2)[0]) + int(str(l[n] * 2)[1])
    else:
        s += 2 * l[n]

# Every other digit starting with last
for n in range(len(l) - 1, -1, -2):
    if l[n] > 9:
        s += int(str(l[n])[0]) + int(str(l[n])[1])
    else:
        s += l[n]

# Check if remainder is 0
if not s % 10:

    # Check for American Express
    if len(l) == 15 and l[0] == 3 and l[1] in [4, 7]:
        print("AMEX")

    # Check for MasterCard
    elif len(l) == 16 and l[0] == 5 and l[1] in [1, 2, 3, 4, 5]:
        print("MASTERCARD")

    # Check for Visa
    elif len(l) in [13, 16] and l[0] == 4:
        print("VISA")

    # Else invalid
    else:
        print("INVALID")

# Else it is invalid
else:
    print("INVALID")

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>

int main(void)
{
    // Get user input
    string input = get_string("Number: ");

    // Get length of input
    int c = 0;
    for (int i = 0; input[i] != '\0'; i++)
    {
        c++;
    }

    // Initialise list of integers
    int number[c];

    // Populate list of integers with digits of input
    for (int i = 0; i < c; i++)
    {
        number[i] =  input[i] - '0';
    }

    // Initialise variable for sum
    int sum = 0;

    // Iterate over every other digit, starting with the last
    for (int i = c - 1; i >= 0; i -= 2)
    {
        sum += number[i];
    }

    // Iterate over every other digit, starting with the secon to last
    for (int i = c - 2; i >= 0; i -= 2)
    {
        if (number[i] * 2 >= 10)
        {
            sum += (number[i] * 2) % 10;
            sum += ((number[i] * 2) / 10) % 10;
        }
        else
        {
            sum += number[i] * 2;
        }
    }

    // Checking validity
    if (!(sum % 10))
    {
        if (c == 15 && number[0] == 3 && (number[1] == 4 || number[1] == 7))
        {
            printf("AMEX\n");
        }

        else if (c == 16 && number[0] == 5 && (number[1] == 1 || number[1] == 2 || number[1] == 3 || number[1] == 4 || number[1] == 5))
        {
            printf("MASTERCARD\n");
        }

        else if ((c == 13 || c == 16) && number[0] == 4)
        {
            printf("VISA\n");
        }

        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
